### ISSUE TITLE
The S3ToGCSOperator fails on templated `dest_gcs` URL

### ISSUE THREAD

**Issue by nvembar** [apache/airflow#14682]
**Apache Airflow version**:

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**:

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**: Docker

**What happened**:

When passing a templatized `dest_gcs` argument to the `S3ToGCSOperator` operator, the DAG fails to import because the constructor attempts to test the validity of the URL before the template has been populated in `execute`.

The error is:

```
Broken DAG: [/opt/airflow/dags/bad_gs_dag.py] Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/gcs.py", line 1051, in gcs_object_is_directory
    _, blob = _parse_gcs_url(bucket)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/google/cloud/hooks/gcs.py", line 1063, in _parse_gcs_url
    raise AirflowException('Please provide a bucket name')
airflow.exceptions.AirflowException: Please provide a bucket name
```

**What you expected to happen**:

The DAG should successfully parse when using a templatized `dest_gcs` value.

**How to reproduce it**:

Instantiating a `S3ToGCSOperator` task with `dest_gcs="{{ var.gcs_url }}"` fails. 

<details>

```python
from airflow.decorators import dag
from airflow.utils.dates import days_ago
from airflow.providers.google.cloud.transfers.s3_to_gcs import S3ToGCSOperator


@dag(
    schedule_interval=None,
    description="Demo S3-to-GS Bug",
    catchup=False,
    start_date=days_ago(1),
)
def demo_bug():

    S3ToGCSOperator(
        task_id="transfer_task",
        bucket="example_bucket",
        prefix="fake/prefix",
        dest_gcs="{{ var.gcs_url }}",
    )


demo_dag = demo_bug()
```
</details>

**Anything else we need to know**:

Should be fixable by moving the code that evaluates whether the URL is a folder to `execute()`.

**Comment by TobKed** [comment]
It may be the same case for `AzureFileShareToGCSOperator`

**Comment by eskarimov** [comment]
I'd like to work on this issue. Seems a similar one might affect other operators as well - @TobKed mentioned [AzureFileShareToGCSOperator](https://github.com/apache/airflow/blob/866a601b76e219b3c043e1dbbc8fb22300866351/airflow/providers/google/cloud/transfers/azure_fileshare_to_gcs.py#L106-L114). [`LocalFilesystemToS3Operator`](https://github.com/apache/airflow/blob/1632c9f519510ff218656bbc1554c80cb158e85a/airflow/providers/amazon/aws/transfers/local_to_s3.py#L97-L98) might work incorrect as well.
Basically any validation of input parameter, which can be templated, shouldn't be inside operator's `__init__` function, as its real value will be available at earliest only during task preparation stage.

How it'd be better to handle? Move validation to the top of `execute()` function?

I'll search then for other potential operators which might be affected
